File: Source/backend/src/TableWorks.Application/Projects/ProjectAutoProgress.cs
namespace ASideNote.Application.Projects;

/// <summary>
/// Linear calendar-day progress between start and end (UTC dates).
/// </summary>
public static class ProjectAutoProgress
{
    public static int ComputePercent(DateTime? startUtc, DateTime? endUtc, DateTime utcNow)
    {
        if (!startUtc.HasValue || !endUtc.HasValue) return 0;

        var start = DateTime.SpecifyKind(startUtc.Value, DateTimeKind.Utc).Date;
        var end = DateTime.SpecifyKind(endUtc.Value, DateTimeKind.Utc).Date;
        var today = utcNow.Kind == DateTimeKind.Utc ? utcNow.Date : utcNow.ToUniversalTime().Date;

        if (today < start) return 0;
        if (today >= end) return 100;

        var totalDays = (end - start).TotalDays;
        if (totalDays <= 0) return 100;

        var elapsed = (today - start).TotalDays;
        var pct = elapsed / totalDays * 100.0;
        return (int)Math.Round(Math.Clamp(pct, 0, 100));
    }
}

File: Source/frontend/src/lib/projectAutoProgress.ts
/**
 * Linear progress (0–100) by calendar day between start and end (ISO date strings, UTC).
 * Matches backend ProjectAutoProgress.ComputePercent.
 */
export function computeProjectAutoProgressPercent(
  startIso: string,
  endIso: string,
  now: Date = new Date(),
): number {
  const start = new Date(startIso);
  const end = new Date(endIso);
  const startDay = Date.UTC(
    start.getUTCFullYear(),
    start.getUTCMonth(),
    start.getUTCDate(),
  );
  const endDay = Date.UTC(end.getUTCFullYear(), end.getUTCMonth(), end.getUTCDate());
  const today = Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate());

  if (today < startDay) return 0;
  if (today >= endDay) return 100;

  const totalMs = endDay - startDay;
  if (totalMs <= 0) return 100;

  const elapsed = today - startDay;
  return Math.round(Math.min(100, Math.max(0, (elapsed / totalMs) * 100)));
}
